File: R/R/l_navgraph.R

#' @title Explore a dataset with the canonical 2d navigation graph setting
#'   
#' @description Creates a navigation graph, a graphswitch, a navigator and a 
#'   geodesic2d context added, and a scatterplot.
#' 
#' @param data a data.frame with numeric variables only
#' @template param_separator
#' @param graph optional, graph or loongraph object with navigation graph. If 
#'   the graph argument is not used then a 3d and 4d transition graph and a 
#'   complete transition graph is added.
#' @param ... arguments passed on to modify the scatterplot plot states
#' 
#' @templateVar page  learn_R_display_graph
#' @templateVar section l_navgraph
#' @template see_l_help
#' 
#' @return named list with \code{graph} handle, \code{plot} handle,
#'   \code{graphswitch} handle, \code{navigator} handle, and \code{context}
#'   handle.
#'   
#' @export
#' 
#' @examples 
#' ng <- l_navgraph(oliveAcids, color=olive$Area)
#' ng2 <- l_navgraph(oliveAcids, separator='-', color=olive$Area)



l_navgraph <- function(data, separator=":", graph=NULL,  ...) {

    if(!is.data.frame(data)) {
        data <- as.data.frame(data)
    }
    
    tt <- l_toplevel()

    if (is.null(graph)) {
        G <- completegraph(nodes=names(data))
        LG <- linegraph(G, sep=separator)
        LGnot <- loon::complement(LG)
        
        cmb <- utils::combn(names(data),2)
        CompG <- completegraph(nodes=apply(cmb[,cmb[1,]!=cmb[2,]],2,
                                   FUN=function(x)paste(x, collapse=separator)))
        
        g <- l_graph(LG, parent=tt)
    } else {
        g <- l_graph(graph, parent=tt)
    }

    tktitle(tt) <- paste0("loon navigation graph: ", g)
    
    gs <- l_graphswitch(activewidget=g, parent=tt)

    if (is.null(graph)) {
        l_graphswitch_add(gs, graph=LG, label="3d transition")
        l_graphswitch_add(gs, graph=LGnot, label="4d transition")
        l_graphswitch_add(gs, graph=CompG, label="3d and 4d transition")
    } else {
        gid <- l_graphswitch_add(gs, graph=graph, label="user specified")
    }

    tkpack(gs, side='right', fill='y')    
    tkpack(g, side='right', fill='both', expand=TRUE)


    navigator <- l_navigator_add(g, label="default navgraph")
    context <- l_context_add_geodesic2d(navigator, data=data, separator=separator)  

    plot <- strsplit(context['command'], " ", fixed = TRUE)[[1]][1]
    class(plot) <- c("l_plot", "loon")

    args <- list(...)
    if (length(args)>0) {
        l_configure(plot, ...)
    }
    
    navgraph <- list(graph=g, plot=plot, graphswitch=gs,
                     navigator=navigator, context=context)
    
    class(navgraph) <- c("l_navgraph", "l_compound", "loon")
    
    navgraph
}

#' @export
l_getPlots.l_navgraph <- function(target){
    list(graph = target$graph, plot = target$plot)
}

File: Python/diver/l_navgraph.py
from .l_toplevel import l_toplevel
from .graphutils import completegraph,linegraph,complement
from .l_graphswitch import l_graphswitch,l_graphswitch_add
from .l_graph import l_graph
from .l_navigator import l_navigator_add
from .l_context2d import l_context_add_geodesic2d
from .loon_class import loon_l_plot, loon_l_navgraph
from .l_configure import l_configure
from .tk import tk
import pandas as pd 
from itertools import combinations
from sys import exit 
def l_navgraph(data, separator=":", graph=None,**options):
    '''Explore a dataset with the canonical 2d navigation graph setting
    
    Creates a navigation graph, a graphswitch, a navigator and a 
    geodesic2d context added, and a scatterplot.

    Args:
        data: a pandas.DataFrame with numeric variables only
        graph: optional, graph or loongraph object with navigation graph. If 
                the graph argument is not used then a 3d and 4d transition graph and a 
                complete transition graph is added.
        **options: arguments passed on to modify the scatterplot plot states
    
    Returns:
        return a loon_l_navgraph object, which include the information of
        `graph` handle, `plot` handle, `graphswitch` handle, `navigator`
        handle, and `context` handle.

    Examples:
        >>> ng = l_navgraph(oliveAcids, color=olive.Area)
        >>> ng2 = l_navgraph(oliveAcids, separator='-', color=olive.Area)
    '''
    if(not isinstance(data,pd.DataFrame)):
        data = pd.DataFrame(data)
    
    tt = l_toplevel()
    if(graph == None):
        G = completegraph(nodes=list(data.columns))
        LG = linegraph(G, separator=separator)
        LGnot = complement(LG)        
        cmb  = list(combinations(list(data.columns),2))
        CompG = completegraph(nodes= [separator.join(x) for x in cmb])
        
        g = l_graph(LG, tt)
    else:
        g = l_graph(graph, tt)
        
    tk.tk.call('wm','title',tt,"loon navigation graph: " + g.plot)
    gs = l_graphswitch(activewidget=g, parent=tt)
    if (graph == None):
        l_graphswitch_add(gs, graph=LG, label="3d transition")
        l_graphswitch_add(gs, graph=LGnot, label="4d transition")
        l_graphswitch_add(gs, graph=CompG, label="3d and 4d transition")
    else:
        gid = l_graphswitch_add(gs, graph=graph, label="user specified")
    tk.tk.call('pack',gs.plot,'-side','right','-fill','y')
    tk.tk.call('pack',g.plot,'-side','right','-fill','both','-expand',True)
    navigator = l_navigator_add(g, label="default navgraph")
    context = l_context_add_geodesic2d(navigator, data=data, separator=separator)  
    plot = context['command'].split(' ')[0]
    plot = loon_l_plot(plot)
    if(len(options) > 0):
        l_configure(plot,**options)

    navgraph = loon_l_navgraph({'graph':g,'plot':plot,'graphswitch':gs,
                                'navigator':navigator,'context':context})

    return navgraph

